core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -240,6 +242,19 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
   }
 
   /**
+    * Remove the watch for the given key and cancel any operations
+    * @param key
+    * @return the delayed operations that were watched
+    */
+  def cancelForKey(key: Any): List[T] = {
```

Comment:
I feel we can instead add another function for directly watching the operation and not "tryCompleteElse..", see comments above.